### ISSUE TITLE
collective OpenMP queue

### ISSUE THREAD

**Issue by psychocoderHPC** [alpaka-group/alpaka#820]
In #783 a workflow is described to use user spawned parallel regions to share the workload between different threads.

example how it can look like
```C++
    // Define the index domain
    using Dim = alpaka::dim::DimInt<1>;
    using Idx = size_t;

    // Define the accelerator
    using Acc = alpaka::acc::AccCpuOmp2Blocks<Dim, Idx>;
    using Dev = alpaka::dev::Dev<Acc>;

    using Queue = alpaka::queue::QueueCpuOmp2Collective;
    using Pltf = alpaka::pltf::Pltf<Dev>;

    auto dev = alpaka::pltf::getDevByIdx<Pltf>(0u);
    Queue queue(dev);

    std::vector<int> results(4, -1);

    using Vec = alpaka::vec::Vec<Dim, Idx>;
    Vec const elementsPerThread(Vec::all(static_cast<Idx>(1)));
    Vec const threadsPerBlock(Vec::all(static_cast<Idx>(1)));
    Vec const blocksPerGrid(results.size());

    using WorkDiv = alpaka::workdiv::WorkDivMembers<Dim, Idx>;
    WorkDiv const workDiv(
        blocksPerGrid,
        threadsPerBlock,
        elementsPerThread);

    // user is spawning a parallel region to controll the number of threads and avoid thread creation during the runtime from alpaka
    #pragma omp parallel num_threads(static_cast<int>(results.size()))
    {
        auto kernel =
        [&] ALPAKA_FN_ACC (
            Acc const & acc) noexcept
        -> void
        {
            size_t threadId = alpaka::idx::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0];
            // avoid that one thread is doing all the work
            std::this_thread::sleep_for(std::chrono::milliseconds(200u * threadId));
            results[threadId] = threadId;
        };

        // the kernel will be executed collectively 
        alpaka::kernel::exec<Acc>(
               queue,
               workDiv,
               kernel);

        alpaka::wait::wait(queue);
    }

    for(size_t i = 0; i < results.size(); ++i)
    {
        REQUIRE(static_cast<int>(i) == results.at(i));
    }
```

**Comment by psychocoderHPC** [comment]
I have implemented it already in alpaka https://github.com/psychocoderHPC/alpaka/tree/topic-collectiveOmpQueue2 but I am not happy with the highly specific queue. 
There are many open questions e.g. 
- How should we handle non kernel tasks? In my current prototype I use `#pragma omp single nowait` for that but maybe the user like to handle it different.
- How should an event be handled? Currently, I perform `#pragma omp barrier`

**Comment by psychocoderHPC** [comment]
We designed alpaka that it can be extended at any time from the user code. We broke this a little bit inside of the CPU device implementation. This avoids that the user can build inside of there own code a collective queue.
I think it is not as hard to fix it and allow the user to implement the queue so there will be no need to introduce a specialized queue.

**Comment by BenjaminW3** [comment]
I had a look at your implementation and was not yet sure if I understand the reasoning behind some design decisions.

What is the expected result if I have `N` OpenMP threads? From reading the code one would think that the N threads each create a kernel function object `kernel`, create a corresponding kernel execution task and then enqueue those tasks which will execute them. So from reading the code it would be the expectation that we would execute the kernel `N` times. If this is not the expectation and we only want the kernel to be executed once at all, than the code does not express this in any way. If this would be the goal I would expect something like this:
```
#pragma omp parallel num_threads(static_cast<int>(results.size()))
{
    #pragma omp single nowait
    {
        auto kernel =
        [&] ALPAKA_FN_ACC (
            Acc const & acc) noexcept
        -> void
        {
            size_t threadId = alpaka::idx::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0];
            // avoid that one thread is doing all the work
            std::this_thread::sleep_for(std::chrono::milliseconds(200u * threadId));
            results[threadId] = threadId;
        };

        // the kernel will be executed collectively 
        alpaka::kernel::exec<Acc>(
               queue,
               workDiv,
               kernel);
    }
}
```

As far as I understand the use case, it is only useful when the queue is Blocking. Else it can not be assured that the task being enqueued is executed within the current parallel region. So it should probably be renamed from `QueueCpuOmp2Collective` to `QueueCpuBlockingOmp2Collective`. 

I would also argue that the user is repsonsible for using the queue only within parallel regions. So we may remove all the `if(::omp_in_parallel() != 0)` and replace it by an assertion or an exception to make the code shorter and easier to understand. However, we can also keep this feature.

**Comment by psychocoderHPC** [comment]
@BenjaminW3 The kernel tasks should be performed collective/cooperative so `#pragma single` avoid it.

I am currently cleaning up my code to allow that such a special queue can be defined in the user code. There are many open questions e.g. who takes care if a memcpy performed in one of the thread is not overlapping a kernel execution. You can always call `ompbarrier`  but this will influence the performance.
IMO such questions are to application-specific to put it in alpaka.

**Comment by sbastrakov** [comment]
@BenjaminW3 I also do not understand your code example. Having the kernel call inside `#pragma omp single` already is a show-stopper for having a collectively executed kernel.

To add up a little bit, I see the idea behind this new collective queue as follows. Using an OpenMP analogy, the difference between this one and the other OpenMP queues is the same as in `#pragma omp for` inside a parallel section vs. `#pragma omp parallel for`: in the first case we are already in a parallel context and enqueue is a work sharing operation (what @psychocoderHPC called collective), while in the second case the enqueue first starts a new parallel context and then does a collective operation inside it.

**Comment by BenjaminW3** [comment]
Yes, my example from above will not work but I wanted make it clear that from reading the code from the topmost example you will expect the kernel will be applied to the workdiv `N` times and not only once. It is absolutely possible to implement something like this, but my problem is that it totally obfuscates what the code really is doing. I would prefer if it would be more visible at the call-site that it is executed only once.

What I really see as problematic from a code standpoint is that many things will be done repeatedly for all threads even though it would only be necessary once. Creating the `kernel` object should only be done once as well as the call to `kernel::createTaskKernel` (part of the convenience `kernel::exec` function). Furthermore, the cal to `kernel::getBlockSharedMemDynSizeBytes` (part of applying the kernel) is executed for each thread instead of only once.

**Comment by BenjaminW3** [comment]
@psychocoderHPC Can this be closed now?

**Comment by BenjaminW3** [comment]
@psychocoderHPC Have you seen my question above?

**Comment by psychocoderHPC** [comment]
> Yes, my example from above will not work but I wanted make it clear that from reading the code from the topmost example you will expect the kernel will be applied to the workdiv `N` times and not only once. It is absolutely possible to implement something like this, but my problem is that it totally obfuscates what the code really is doing. I would prefer if it would be more visible at the call-site that it is executed only once.

I agree that the concepts of the queue are not clearly visible in the code. Therefore #822 moved the implementation of this highly special queue to the user.

> What I really see as problematic from a code standpoint is that many things will be done repeatedly for all threads even though it would only be necessary once. Creating the `kernel` object should only be done once as well as the call to `kernel::createTaskKernel` (part of the convenience `kernel::exec` function). 

I do not see any problems that each thread is creating the kernel object. A collective creation would be much slower because it requires some kind of synchronization or communication between the threads. Since the calls will be mostly from OpenMP threads those are inside a parallel section the created object will private (good for performance)

> Furthermore, the cal to `kernel::getBlockSharedMemDynSizeBytes` (part of applying the kernel) is executed for each thread instead of only once.

I do not see this as big issue if the shared memory calculation is not depending on a state that is increased with each kernel call.

**Comment by psychocoderHPC** [comment]
I will close this issue since the needed queue implemented in #822 and can be used and adjusted by the user.
